cmd/settlement_gemini.go

```diff
@@ -0,0 +1,297 @@
+func geminiMapTransactionsToID(transactions []settlement.AntifraudTransaction) map[string]settlement.Transaction {
+	transactionsMap := make(map[string]settlement.Transaction)
+	for _, atx := range transactions {
+		tx := atx.ToTransaction()
+		transactionsMap[gemini.GenerateTxRef(&tx)] = tx
+	}
+	return transactionsMap
+}
+
+// GeminiWriteTransactions writes settlement transactions to a json file
+func GeminiWriteTransactions(outPath string, metadata *[]settlement.Transaction) error {
+	if len(*metadata) == 0 {
+		return nil
+	}
+	data, err := json.MarshalIndent(metadata, "", "  ")
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(outPath, data, 0600)
+}
+
+// GeminiWriteRequests writes settlement transactions to a json file
+func GeminiWriteRequests(outPath string, metadata *[][]gemini.PayoutPayload) error {
+	data, err := json.MarshalIndent(metadata, "", "  ")
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(outPath, data, 0600)
+}
+
+// GeminiTransformForMassPay starts the process to transform a settlement into a mass pay csv
+func GeminiTransformForMassPay(input string, oauthClientID string, output string) (err error) {
+	transactions, err := settlement.ReadFiles(strings.Split(input, ","))
+	if err != nil {
+		return err
+	}
+
+	geminiPayouts, err := GeminiTransformTransactions(oauthClientID, *transactions)
+	if err != nil {
+		return err
+	}
+	err = GeminiWriteRequests(output, geminiPayouts)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GeminiConvertTransactionsToGeminiPayouts converts transactions from antifraud to "payouts" for gemini
+func GeminiConvertTransactionsToGeminiPayouts(transactions *[]settlement.Transaction, txID string) (*[]gemini.PayoutPayload, decimal.Decimal) {
+	payouts := make([]gemini.PayoutPayload, 0)
+	total := decimal.NewFromFloat(0)
+	for i, tx := range *transactions {
+		tx.SettlementID = txID
+		(*transactions)[i] = tx
+		payout := gemini.SettlementTransactionToPayoutPayload(&tx)
+		total = total.Add(payout.Amount)
+		payouts = append(payouts, payout)
+	}
+	return &payouts, total
+}
+
+// GeminiTransformTransactions splits the transactions into appropriately sized blocks for signing
+func GeminiTransformTransactions(oauthClientID string, transactions []settlement.Transaction) (*[][]gemini.PayoutPayload, error) {
+	maxCount := 500
+	blocksCount := (len(transactions) / maxCount) + 1
+	privateRequests := make([][]gemini.PayoutPayload, 0)
+	i := 0
+
+	txnID := transactions[0].SettlementID
+	txID := uuid.Must(uuid.FromString(txnID))
+	fmt.Println("transaction id", txID.String())
+
+	fmt.Printf("creating %d blocks\n", blocksCount)
+	fmt.Printf("with %d transactions\n", len(transactions))
+	total := decimal.NewFromFloat(0)
+	for i < blocksCount {
+		// var transactionBlock []settlement.Transaction
+		lowerBound := i * maxCount
+		upperBound := (i + 1) * maxCount
+		payoutLength := len(transactions)
+		if payoutLength <= upperBound {
+			upperBound = payoutLength
+		}
```

Comment:
Is this code also in other places? I was having a little bit of working through if we could get into a case where we're missing an item.

Reply:
tested manually and worked. this code has to do with chopping off the upper bound of an index counter such that on a series of 1900 transactions that have to be blocked by 500, we do not try to access index 2000